---
title: "Genomic Heritability"
author: "Malachy Campbell"
date: "10/16/2018"
header-includes:
   - \usepackage{bbm}
   - \usepackage{amsmath}
output: 
  beamer_presentation:
  theme: "CambridgeUS"
  colortheme: "beaver"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Learning Objectives

- Understanding heritability
- Quantify relatedness
    - Identity by descent (IBD) and Identity by state (IBS)
    - Pedigree-based approach (brief): A matrix 
    - Marker-based approach: G matrix
- Estimating genetic variances via the "Animal" model (brief)
- Cautionary tale of genomic heritability

## What is heritability?

## What is heritability?

- Proportion of total phenotypic variance in a population explained by total genetic ($H^2$) or additive genetic effects ($h^2$)
$$Y = G + E$$
 - In the broad sense:
    $$ V_Y = V_A + V_D + V_I + V_E $$
    $$ H^2 = \frac{V_A + V_D + V_I}{V_A + V_D + V_I + V_E} $$
- In the narrow sense:    
    $$ h^2 = \frac{V_A}{V_A + V_D + V_I + V_E} $$

## Genetic values (a)
- Genetic values (a) are a linear function of allele content at **QTL** 
$$y = a + e = \alpha'z_i + e$$
- Thus $a = \alpha'z_i$, where $\alpha'$ is a vector of additive effects at QTL and $z'$ is a vector of genotypes at each QTL 
 <!---
 The true additive genetic value for an individual is the summation of QTL effects for its genotype. This is often unknown, so we can approximate it with markers
 --->
 
- $h^2$ is the proportion of phenotypic variation that can be explained by a regression of phenotypes on QTL

## Genomic values (g)
- Genomic values (g) are a linear function of allele content at **markers**

- Thus $g = \beta'x_i$, where $\beta'$ is a vector of additive effects at markers and $x'$ is a vector of genotypes at each marker

- Genomic heritability ($h_g^2$) is the proportion of phenotypic variation that can be explained by a regression of phenotypes on markers

## Relationship between $h^2$ and $h_g^2$
- Genomic values ($g$) are only an approximation of genetic values ($a$)
$$\alpha'z_i = \beta'x_i + e$$
$$a = g + \bar{g}$$

- The true genetic value ($a$) is the genomic value ($g$) plus some genetic effects ($\bar g$) that cannot be captured by markers 
    - $\bar g$ is the 'missing heritability'

## How do we measure heritability?

- Prior to genetic markers what did we need?

## How do we measure heritability?

- What do we need?

    - Need a population with related individuals
 <!---
 Rationale is that parents pass on genetic information to their offspring. Various mating designs allow $h^2$ to be quantified. If we have a complex popualtion we need some way to quantify relationships via pedigree
 --->
    
    - Need a means to assess genetic relatedness between individuals
<!---
 If we have a complex popualtion we need some way to quantify relationships via pedigree. Now with markers we can assess relationships without knowning the pedigree.
 --->
    
    - Need methods to partition variance into genetic and non-genetic effects
<!---
Regression: Regress offspring on parental phenotypes
ANOVA: Partition variance into within and among family variance
ML/REML: Leverage genetic relatedness in complex populations to partition phenotypic variance to genetic and non-genetic components (can accomidatate complex genetic realtionships)
--->

## Estimating genetic relatedness via pedigrees
    
- **Coefficient of ancestry (kinship coefficient)**: What is the \textcolor{red}{probability} that two alleles at the same locus sampled at random from two individuals are derived from a common ancestor (e.g. are Identical By Descent, IBD)

    - Two randomly sampled alleles at a locus are IBD if they have the same ancestral origin.

## Estimating genetic relatedness via pedigrees: expected relatedness

- The expected relatedness between individuals is twice the kinship coefficient

    - This expected relatedness matrix (A-matrix) represents the expected additive genetic relationships between individuals in a population

## Marker-based approach

- **Goal**: Derive a relationship matrix, like $\bf A$, that represents the **realized** genetic similarities between individuals using genetic markers
    - Genomic realtionship matrix ($\bf G$)

1. Determine the proportion of chromosome segments shared via the **identical by state (IBS)** matching of marker alleles.
2. Scale markers to more closely reflect **IBD** relationships

## Tiny GRM example
- ${\bf M}$: $n x m$ matrix of markers (aa = 0, Aa = 1, AA = 2)
```{r, echo = F}
M <- matrix(c(0,1,0,2,
              2,1,1,1,
              2,0,0,0), 3, 4, byrow = T)

print(M)
```  
- ${\bf W}$: subtract 1 to rescale ${\bf M}$ to -1, 0, 1
```{r, echo = F}
M <- matrix(c(0,1,0,2,
              2,1,1,1,
              2,0,0,0), 3, 4, byrow = T)

print(M-1)
```  
## Tiny GRM example: **identical by state (IBS)** matching of marker alleles
- Determine the homozygous identity in state matching by taking cross product (${\bf WW'}$)
    - Diagonal = # of homozygous loci in each individual
    - Off-diagonal = (# of loci with matching homozygous genotypes) - (# number non-matching homozygous loci)
```{r, echo = F}
M <- matrix(c(0,1,0,2,
              2,1,1,1,
              2,0,0,0), 3, 4, byrow = T)

print((M-1) %*% t(M-1))
```     

- \textcolor{red}{${\bf WW'}$ is an $n x n$ {\bf IBS} similarity matrix}


## Marker-based approach: Rescale to reflect **identical by state (IBD)**
- In the previous example common and rare alleles have the same weight on genomic relatedness between individuals
- If two individuals share a rare allele there should be a greater chance that they are closely related
    - Therefore for each marker $i$ in ${\bf M}$, center the marker scores by the mean marker score ($2 \hat{p_i}$) where $p_i$ is the minor allele frequency (MAF) of marker $i$

## Marker-based approach: Rescale to reflect **identical by state (IBD)**
- Suppose the four markers have a MAFs of 0.01, 0.15, 0.25, and 0.5. Subtracting ($2 \hat{p_i}$) from ${\bf M}$ will give us ${\bf Z}$
- ${\bf ZZ'}$
```{r, echo = F}
M <- matrix(c(0,1,0,2,
              2,1,1,1,
              2,0,0,0), 3, 4, byrow = T)

maf <- matrix(c(rep(c(0.01, 0.15, 0.25, 0.5), 3) ), 
              3, 4, byrow = T)

Z <- M - 2*maf
print(Z %*% t(Z))
```     

- ${\bf WW'}$
```{r, echo = F}
M <- matrix(c(0,1,0,2,
              2,1,1,1,
              2,0,0,0), 3, 4, byrow = T)

maf <- matrix(c(rep(c(0.01, 0.15, 0.25, 0.5), 3) ), 
              3, 4, byrow = T)

print((M-1) %*% t(M-1))
``` 

## Marker-based approach: scale ${\bf G}$ so that it is analogous to ${\bf A}$
-  As the number of markers in ${\bf Z}$ increases, so does the elements of ${\bf ZZ'}$
    - To be comparable we must make ${\bf ZZ'}$ independant of the number of markers
    - Dividing by the sum of variances at each locus $2\sum_{1}^{i}p_i(1-pi)$ gives us ${\bf G}$ a **realized** relationship matrix that has the similar properties to ${\bf A}$
    
    $$\mathbf{G} = \frac{\mathbf{ZZ'}}{2\sum_{1}^{i}p_i(1-pi)}$$
```{r, echo = F}
M <- matrix(c(0,1,0,2,
              2,1,1,1,
              2,0,0,0), 3, 4, byrow = T)

maf <- c(0.01, 0.15, 0.25, 0.5)

Z <- M - 2*maf
sumvar <- 2*sum(maf*(1-maf))
print((Z %*% t(Z)) / sumvar)
```  
    
## Genomic relationship matrix ${\bf G}$
- Elements of ${\bf G}$ are twice the realized kinship coefficients
- Diagonal elements indicate the degree of inbreeding for individuals ($E(\Theta_{ii}) = 1 - F_{i}$, where $F_{i}$ is the inbreeding coefficient)
    - 1 for non-inbred individuals
```{r, echo = F}
M <- matrix(c(0,1,0,2,
              2,1,1,1,
              2,0,0,0), 3, 4, byrow = T)

maf <- c(0.01, 0.15, 0.25, 0.5)

Z <- M - 2*maf
sumvar <- 2*sum(maf*(1-maf))
print((Z %*% t(Z)) / sumvar)
```

## Estimating genetic parameters with $\mathbf{G}$ via the "Animal" model

$$\mathbf{y} = \mathbf{X\boldsymbol{\beta}} + \mathbf{Zu} + e$$

- $\mathbf{y}$ ($n \times 1$) vector of observations

- $\boldsymbol{\beta}$ ($p \times 1$) vector of fixed effects (year, location, etc.)

- $\mathbf{u}$ ($q \times 1$) vector of genetic values
    - $q$ is all the individuals in $\mathbf{G}$, $q > n$
    - $\mathbf{u} \sim N(0, \mathbf{G}\sigma_g^2$), \textcolor{red}{covariance in genetic values follows from genetic covariance between individuals}

-  $e$ residual effects 
    - $e \sim N(0, \mathbf{I}_n \sigma_e^2)$

## Estimating genetic parameters with $\mathbf{G}$ via the "Animal" model

\begin{align}
\begin{bmatrix}
\mathbf{X'X}  & \mathbf{X'Z} \\
\mathbf{Z'X}  & \mathbf{Z'Z} + \lambda \mathbf{G^{-1}} \\
\end{bmatrix} 
\begin{bmatrix}
\boldsymbol{\hat\beta} \\
\boldsymbol{\hat u} \\
\end{bmatrix}
=
\begin{bmatrix}
\boldsymbol{X'y} \\
\boldsymbol{Z'y} 
\end{bmatrix}
\end{align}

\begin{align}
\lambda = \frac{\sigma_e^2}{\sigma_g^2} = \frac{1 -h^2}{h^2}
\end{align}

## Estimating genetic parameters with $\mathbf{G}$ via the "Animal" model
- Use maximum likelihood (ML) or restricted maximum likelihood to solve model
    - Goal is to find a set of parameters that maximizes the likelihood of the data
    - ML: Estimates all parameters together; assumes no error in estimating fixed effects
    - REML: Allows for loss of degrees of freedom for estimating fixed effects
    
- Little difference between ML or REML if # of fixed effects is small
- No need to solve it by hand!
    - standalone: ASREML, GCTA
    - R: 'nlme', 'lme4', 'ASREML-R'

## A cautionary tale of genomic heritability

- Is genomic heritability a good approximation of $h^2$?

- $\mathbf{G} = \frac{\mathbf{ZZ'}}{2\sum_{1}^{i}p_i(1-pi)}$

    - What information is captured with $\bf Z$?
    
## A cautionary tale of genomic heritability

- Is genomic heritability a good approximation of $h^2$?

- $\mathbf{G} = \frac{\mathbf{ZZ'}}{2\sum_{1}^{i}p_i(1-pi)}$
    
    - What information is captured with $\bf Z$?
        
        - QTLs are typed by markers
        
        - Markers in LD with QTL
        
        - Markers in LE with QTL

<!---
The real genetic values depend only on QTL, not additional 'non-pertent' markers
--->

## A cautionary tale of genomic heritability: de los Campos, Sorrensen, Gianola (2015)

- Six simulated scenarios for construction of $\bf G$: 1. Only QTL; 2. QTL and markers in LD with QTL; 3. All loci (causal and non-causal); 4. Only markers in LD with QTL; 5. Only markers in LE with QTL (non-informative for $\sigma_p^2$); 6. All markers (LD and LE with QTL)
    - $h^2$ fixed at 0.5

![dlC](delosCampos_2015.jpg){width=225px}    

<!---
- Scenarios 1-3 expect no missing heriability because QTLs are included

- Scenarios 4-6 expect missing heriability because QTLs not included; 5 is essentially the null model
--->

- \textcolor{red}{Incomplete linkage between causal loci and marker leads to missing heritability. Thus $h^2 \neq h_g^2$}


## A cautionary tale of genomic heritability: de los Campos, Sorrensen, Gianola (2015)

- Comparing $h^2$ and $h_g^2$ in related (FHS) and unrelated populations (GEN)

![dlC2](delosCampos_2013.jpg){width=250px}

- Proportion of allele sharing at markers and QTL are greater for related individuals compared to unrelated individuals

    - \textcolor{red}{Cosegregation of markers and QTL are due to recent relationships}

