---
title: "Estimating genomic breeding values using genomic BLUP and ridge regression BLUP"
author: "Malachy Campbell"
date: "11/13/2018"
header-includes:
   - \usepackage{bbm}
   - \usepackage{amsmath}
output: 
  beamer_presentation:
  theme: "CambridgeUS"
  colortheme: "beaver"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, size = "footnotesize")
library(rrBLUP)
```

# Intro

$$ \mathbf{y} = \mathbf{g} + \epsilon $$
- What is $\textbf{g}$?

\begin{table}
\begin{tabular}{l | c | c | c }
Plant ID & y & g & e\\
\hline \hline
1 & 10 & 5 & 5 \\ 
2 & 7 & 6 & 1 \\
3 & 12 & 2 & 10 \\
\end{tabular}
\end{table}

<!--- Which line is best? If we select only on phenotypic data, which would you choose? Phenotypic selection is unreliable. The goal is to predict g with some reliable accuracy. Like I talked about earlier, BLUP was traditionally done with pedigree data, but if we don't have pedigree data, we need to use the next best alternative - markers
--->

# Intro

- Recall that $\mathbf{g}$ is the cumulative additive genetic effect 

$$ \mathbf{y} = 1 \mu + \sum_k x_k \beta + e $$
$\mathbf{W}$ is a centered $n \times m$ marker matrix, $\mathbf{a}$ is vector of SNP effects 

<!--- 
Next logical step would be if we want to predict g, we can just predict all the SNP effects
--->


# Ridge regression BLUP

$$ \mathbf{y} = 1 \mu + \sum_k x_k \beta + \epsilon $$

- Proposed before the 'big data' trend by Meuwissen et al (2001)

- $\hat{\beta} = \mathbf{X'}(\mathbf{X'X} + \lambda \mathbf{I})^{-1} \mathbf{y}$

    - $\lambda = \frac{\sigma^2_e}{\sigma^2_\beta}$

# Genomic BLUP

$$ \mathbf{y} = 1 \mu + \mathbf{Zu} + e $$

- $\hat{u} = \left[ \mathbf{I} + \mathbf{G}^{-1} \frac{\sigma^2_e}{\sigma^2_u} \right] \mathbf{y}$

- How do these methods differ? How will the GEBVs differ?

# Equivalence

- For gBLUP the $Var(y) =  \mathbf{ZGZ'}\sigma^2_u + \mathbf{I}\sigma^2_e$

- For rrBLUP the $Var(y) =  \mathbf{XX'}\sigma^2_\beta + \mathbf{I}\sigma^2_e$ 
    
    - What does $\mathbf{XX'}$ represent?

# Demonstration with Spindel data

- 299 elite rice lines from IRRI 

- genotyped with 73,147 SNPs

    - we'll use 39,560
    
- phenotyped for 19 traits 

    - **grain yield (GY)**
    
    - measured in dry and wet seasons

![](spindel.png){width=225px}

# Loading data

\footnotesize
```{r load spindel data, echo = T}
##Clear all objects
rm(list = ls())

# Load the data
pheno <- read.csv("~/Downloads/Spindel/pheno_WS.csv")
dim(pheno)

geno <- read.table("~/Downloads/Spindel/Spindel_geno.txt", 
                   sep = "\t", header = T, row.names = 1)
dim(geno)
geno <- t(geno)
dim(geno)

sum(row.names(geno) == pheno$GHID)
```

# Calculate a GRM
\footnotesize
```{r GRM for spindel, echo = T}
head(geno[,1:5])

Zsc <- scale(x = geno, center = T, scale = T)
GRM <- tcrossprod(Zsc)/ncol(geno)

dim(GRM)
```


# gBLUP using rrBLUP package
\footnotesize
```{r gBLUP via rrBLUP, echo = T}
library(rrBLUP)

gBLUP <- mixed.solve(y = pheno$YLD, K = GRM)
names(gBLUP)

length(gBLUP$u)
```


# rrBLUP using rrBLUP package
\footnotesize
```{r rrBLUP via rrBLUP, echo = T}
library(rrBLUP)

rrBLUP <- mixed.solve(y = pheno$YLD, Z = Zsc)
names(rrBLUP)

length(rrBLUP$u)
```

- Why are the sizes `rrBLUP$u` and `gBLUP$u` different?

- How can we make the two comparable?

# Are rrBLUP and gBLUP equivalent?

- Recall $$\hat g = W \hat a$$ 


\footnotesize
```{r rrBLUP v gBLUP, echo = T}
#calculate GEBVs from predicted marker effects
gBLUP_rr <- Zsc %*% rrBLUP$u

gBLUP_YLD <- gBLUP$u + as.numeric(gBLUP$beta)
gBLUP_rr_YLD <- gBLUP_rr + as.numeric(rrBLUP$beta)
```

# Are rrBLUP and gBLUP equivalent?
\footnotesize
```{r rrBLUP v gBLUP plot, echo = T, fig.height=1.7, fig.width=3.2, fig.align="center"}
par(mar=c(3,4,0.5,0.5), mgp=c(1.8,0.5,0), xpd = F, cex.lab = 0.5, 
    cex.axis = 0.5)
plot(gBLUP_YLD, gBLUP_rr_YLD, ylab = "Predicted YLD (RR-BLUP)", 
     xlab = "Predicted YLD (gBLUP)", pch = 21, cex = 0.5)

abline(lm(gBLUP_rr_YLD ~ gBLUP_YLD), col = "red")

text(x = 4400, y = 5200, paste0("r = ", 
  round(cor(gBLUP_YLD, gBLUP_rr_YLD),2)), col = "red", cex = 0.75)
```

# How accurate are our predictions?

- How can we estimate how accurate our predicted genomic breeding values are?

# How accurate are our predictions?

- How can we estimate how accurate our predicted genomic breeding values are?

    - Compare predicted and observed breeding values for a new population

    - Partition dataset and use one for training and one for prediction

# Two fold cross validation

- For some dataset
    (1) randomly split the the individuals into two equal sized (or close to) sets
    (2) mask the observations in one set (testing set), keep observations for other set (training set)
    (3) fit the model using training set and predict the values for the missing individuals
    (4) take the correlation between predicted GEBVs for test set and observed phenotypes for test set
    (5) repeat 1 - 4
    
# Two fold cross validation
```{r CV data, echo = T}
pheno_train <- pheno
#define the testing and training sets
set.seed(123)
train_set <- sample (1:length(pheno$GHID), size = length(pheno$GHID)/2)
test_set <- setdiff(1:length(pheno$GHID), train_set)
length(train_set)
length(test_set)
#Mask the phenotypes for the testing set
pheno_train[test_set ,]$YLD <- NA
```

# Run RRBLUP with training set
```{r training fit, echo = T}
library(rrBLUP)
##rrBLUP
rrBLUP_train <- mixed.solve(y = pheno_train$YLD, Z = Zsc)
rrBLUP_train <- Zsc %*% rrBLUP_train$u
length(rrBLUP_train)
```

# Assess predictive ability from rrBLUP approaches
```{r PA, echo = T}
rrBLUP_test <- rrBLUP_train[test_set]
pheno_test <- pheno[test_set ,]

cor(pheno_test$YLD, rrBLUP_test)
```

# References

- Endelman, J. B. Ridge regression and other kernels for genomic selection with R package rrBLUP. Plant Genome 4, 250–255 (2011).

- Habier, D., Fernando, R. L. & Dekkers, J. C. M. The impact of genetic relationship information on genome-assisted breeding values. Genetics 177, 2389–2397 (2007).

- Meuwissen, T. H. E., Hayes, B. J. & Goddard, M. E. Prediction of total genetic value using genome-wide dense marker maps. Genetics 157, 1819–1829 (2001).

- Spindel, J. et al. Genomic selection and association mapping in rice (Oryza sativa): effect of trait genetic architecture, training population composition, marker number and statistical model on accuracy of rice genomic selection in elite, tropical rice breeding lines. PLoS Genet. 11, e1004982 (2015).
